src/main.cpp

```diff
@@ -862,16 +870,45 @@ bool CheckTransaction(const CTransaction& tx, CValidationState &state,
     }
 }
 
-bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidationState &state)
+bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidationState &state, uint32_t consensusBranchId)
 {
     // Basic checks that don't depend on any context
 
-    // Check transaction version
-    if (tx.nVersion < MIN_TX_VERSION) {
+    // Check min and max tx version of pre-Overwinter and Overwintered transactions.
+    if (tx.fOverwintered) {
+        if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
+            return state.DoS(100, error("CheckTransaction(): unknown overwintered version"),
+                REJECT_INVALID, "unknown-overwintered-version");
+        }
+    }
+    else if (tx.nVersion < MIN_TX_VERSION) {
         return state.DoS(100, error("CheckTransaction(): version too low"),
                          REJECT_INVALID, "bad-txns-version-too-low");
     }
 
+    // When checking the transaction against the consensus rules for the Overwinter branch, any transaction
+    // with version format >= 3 must have the fOverwintered flag set.
+    // With Sprout consensus rules, where the branch id is 0, we do not need to perform this check as the
+    // Sprout epoch allows for transactions with version number >= 3 (without setting fOverwintered flag).
+    if (NetworkUpgradeInfo[Consensus::UPGRADE_OVERWINTER].nBranchId == consensusBranchId) {
```

Comment:
I was going to comment that this might be better as a contextual rule, since we want it to apply for all future upgrades. However, once Overwinter is active, if we have never seen a transaction pre-Overwinter that violates this rule, we could remove this gate and just enforce it everywhere, in which case it requires no context and makes sense here. So we can leave it as-is.

Reply:
Yep, I spent a while fiddling with this last night, and I think this works for now.